### PR TITLE
Add 12 num::NonZero* types for primitive integers, deprecate core::nonzero

### PR DESCRIPTION
**Pull request by SimonSapin**
RFC: https://github.com/rust-lang/rfcs/pull/2307
Tracking issue: ~~https://github.com/rust-lang/rust/issues/27730~~ https://github.com/rust-lang/rust/issues/49137
Fixes https://github.com/rust-lang/rust/issues/27730

### CODE REVIEW

**Comment by kennytm** [comment]
Marking as blocked by rust-lang/rfcs#2307, the RFC has not been accepted yet.

File: src/test/rustdoc/synthetic_auto/no-redundancy.rs [hunk 1]
```diff
@@ -20,7 +20,7 @@ where
 
 // @has no_redundancy/struct.Outer.html
 // @has - '//*[@id="synthetic-implementations-list"]/*[@class="impl"]/*/code' "impl<T> Send for \
-// Outer<T> where T: Copy + Send"
+// Outer<T> where T: Send + Copy"
```
Comment (SimonSapin) [inline src/test/rustdoc/synthetic_auto/no-redundancy.rs]: The `ptr::Shared` removal makes this test fail unless I make this change. I have no idea why.

Comment (SimonSapin) [reply src/test/rustdoc/synthetic_auto/no-redundancy.rs]: Filed https://github.com/rust-lang/rust/issues/49123

**Comment by SimonSapin** [comment]
https://github.com/rust-lang/rfcs/pull/2307 has been accepted. New tracking issue: https://github.com/rust-lang/rust/issues/49137

**Comment by KodrAus** [comment]
I was just waiting for the FCP to finish :) Will take a look through shortly.

File: src/libcore/nonzero.rs [hunk 2]
```diff
@@ -62,14 +63,11 @@ impl_zeroable_for_integer_types! {
 /// NULL or 0 that might allow certain optimizations.
 #[lang = "non_zero"]
 #[derive(Copy, Clone, Eq, PartialEq, Ord, PartialOrd, Debug, Hash)]
-pub struct NonZero<T: Zeroable>(T);
+pub struct NonZero<T: Zeroable>(pub(crate) T);
```
Comment (KodrAus) [inline src/libcore/nonzero.rs]: Is there a reason we need to make this `pub(crate)` instead of using the `new_unchecked` and `get` methods?

Comment (SimonSapin) [reply src/libcore/nonzero.rs]: This is in anticipation for later making this module private and reducing it to its bare minimum (the lang item) https://github.com/SimonSapin/rust/pull/1

Comment (SimonSapin) [reply src/libcore/nonzero.rs]: https://github.com/SimonSapin/rust/blob/private-core-nonzero/src/libcore/nonzero.rs

Comment (KodrAus) [reply src/libcore/nonzero.rs]: Ah gotcha. I missed that part in the RFC :+1:

**Review by KodrAus — APPROVED** [review]
This looks good to me! Thanks @SimonSapin

**Comment by alexcrichton** [comment]
@bors: retry

* network error?

**Pull request merged**